### PR TITLE
Remote Projectiles

### PR DESCRIPTION
**Pull request by Fate-JH**
I shot a missile into the air,
Guided across Auraxis, exactly where;
For, so drunkenly it flew, the enemy jock
Could not escape its automatic lock.

Normal projectiles, when having been fired by one player but seen on another player's client, exist as a holographic illusion of combat.  Their form is defined by the fake ammunition loaded into the fake weapon, _whatever that is_.  They appear to interact with the game world properly but really only serve to communicate the general impression of weapon discharge.  The discharge is not even accurate to the place where it targets or strikes.

Some projectiles lack this fake nature and do not render properly on others' clients on their own even if all of the usual packets are traded. These "remote projectiles" require nearly the same level of synchronized representation as another player object or a vehicle.  They are explicitly constructed on each client other than the original shooter's and, by constant report of the client that dispatched the missile, guided to their intended destination on everyone's client.  Once their damage is rendered, they must be safely destroyed and their global identification number unlisted, no different from the aforementioned player object or vehicle.  Most of these projectiles also interact with their targets in an cautionary audible way, through awareness of their operating client's targeting reticle.

___Features___
__`ProjectileDefinition`__
This file actually already existed.  Three specific fields are defined for the purposes of remote projectile construction.

__`RemoteProjectileData`__
This file actually already existed as "Tracked projectileData" but has been renamed, as noted.  More than just the name, however, it was adjusted internally.  The internal structure that held data for the packet and differentiated between different kinds of remote projectiles has been distributed in a different form in `GlobalDefinitions`.  The data herein is retained for archival reasons.  How projectiles actually translate to their packet is still more or less a mystery.

__`ProjectileConverter`__
The converter for `Projectile` objects to `RemoteProjectileData` constructor packet data.

___Caveats___
1. The Vanu Starfire, though functional, is terribly weak, probably owing to its reliance on the aggravated damage status/type.
2. The Decimator (`phoenix`) is configured to use an alternate fire mode setup that is not standard.  Instead of letting the fire cycles of `ChangeFireState_Start` and `ChangeFireState_Stop` play out normally, the response to those packets are overrode to produce a remote projectile that bears the same characteristics of the camera missile (`phoenix_missile_guided_projectile`).  When the projectile causes `SplashHitDamage` by exploding against a target or being set to explode early by the user, and when it is cleared up, a special consideration is made to ensure the projectile displays exactly where the explosion should occur.  This lack of controllable detonation is the main reason everything here has been enacted.  I'm frankly surprised it even manually spawns the projectile in the first place!  It is not a remote projectile found anywhere in our packet captures and its data has been manually approximated such that its maximum distance matches against a standard Decimator projectile.  It does work, however.
3. The Scorpion (`oicw`) has not been worked on beyond basic and common features at the object definition level and should still be relatively ineffective.  Bluntly, it should not work.  The existence of the "oicw little buddies" makes for a complicated second stage projectile interaction that is uncommon to the game engine.  Only a handful of weapons demonstrate an effect approaching this functionality.
4. `Meteors` are also remote projectiles but they have not been worked on either.  If unconditional devastating death is to rain down on a battlefield, anyway, it will take the form of a Suppressor before it will a console command.
5. The original working method of making projectiles animate - creating the projectile anew for each step that it was to be animated as indicated by `ProjectileStateMessage` - has been retained in a comment in `WorldSessionActor`'s `AvatarResponse.ProjectileState` case statement.  The operation is just an `ObjectCreateMessage` call.  By comparison to the natural method of projectile control, in my opinion, this method renders the projectile and its route much closer to where it actually appears on the controller's client.  The path is straighter.  Swap commented and uncommented sections and watch from a third party to test.

___Addenda___
1. I left the `sbt-unidocs` plugin "turned off".  It never worked for me and hindered my ability to start the server and always had it commented while working.
2. In the course of working on remote projectiles, I worked on the `Codec` for the `DamageMessage` packet and the `DamageFeedbackMessage` packet.  They were of no help.
3. Despite how many times I have correct "lootedP" in this branch, the typo never goes away.

### CODE REVIEW

File: pslogin/src/main/scala/WorldSessionActor.scala [hunk 1]
```diff
@@ -4265,9 +4322,15 @@ class WorldSessionActor extends Actor with MDCContextAware {
       }
 
     case msg @ ChangeFireStateMessage_Start(item_guid) =>
-      log.info("ChangeFireState_Start: " + msg)
+      log.trace("ChangeFireState_Start: " + msg)
       if(shooting.isEmpty) {
         FindEquipment match {
+          //special case - suppress the decimator's alternate fire mode
+          case Some(tool : Tool)
```
Comment (SouNourS) [inline pslogin/src/main/scala/WorldSessionActor.scala]: line 4329 & 4334, same case, is that normal ?

Comment (Fate-JH) [reply pslogin/src/main/scala/WorldSessionActor.scala]: It's not the same case.  This case is sibject to the guard conditionals on lines 4330 and 4331.

Comment (Mazo) [reply pslogin/src/main/scala/WorldSessionActor.scala]: I think @SouNourS means you have `case Some(tool : Tool)` twice?

The two could potentially be merged and if the guard condition for the phoenix is met then set `shooting = Some(item_guid)` but the rest can be shared?

File: pslogin/src/main/scala/WorldSessionActor.scala [hunk 2]
```diff
@@ -5416,7 +5488,7 @@ class WorldSessionActor extends Actor with MDCContextAware {
             }
             val (angle, attribution, acceptableDistanceToOwner) = obj match {
               case p : Player =>
-                (p.Orientation, tool.Definition.ObjectId, 10f) //TODO upper body facing
+                (p.Orientation, tool.Definition.ObjectId, 10f + (if(p.Velocity.nonEmpty) { 5f } else { 0f })) //TODO upper body facing
```
[the remark this answers is not in the document]
Comment (Fate-JH) [reply pslogin/src/main/scala/WorldSessionActor.scala]: It's unfortunately not that simple, not at the moment. The range of `FacingYawUpper ` is incorrect for this kind of simple math.  It's one of a few things I've been meaning to fix about angles for a while.

[the remark this answers is not in the document]
Comment (Mazo) [reply pslogin/src/main/scala/WorldSessionActor.scala]: Huh interesting, when I first messed around with that projectiles branch it seemed to be close enough.

**Comment by Fate-JH** [comment]
Update to fix turret upgrade process.
It's, like, two lines in `GlobalDefinitions`.

File: pslogin/src/main/scala/WorldSessionActor.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fate-JH) [reply pslogin/src/main/scala/WorldSessionActor.scala]: This suggestion will be incorporated into a future PR.

**Comment by Fate-JH** [comment]
Short-lived projectiles should not cause (as many) issues.

**Comment by Fate-JH** [comment]
I merged changes using the web interface.
Reminder that: if I perform any more updates to this branch before I merge with master, I need to pull the branch into my local repo to make certain updates are correct.

**Comment by Fate-JH** [comment]
I decided to pre-emptively look at the code for implants and stamina.

**Pull request merged**